packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -14,6 +14,46 @@ import 'package:vector_math/vector_math_64.dart';
 
 import 'debug.dart';
 
+/// Data collected during a layer hit test.
+class LayerHitTestEntry<T> {
+  /// Creates a hit test entry.
+  LayerHitTestEntry(this.value);
+
+  /// The [HitTestTarget] encountered during the hit test.
+  final T value;
+
+  @override
+  String toString() => '$value';
+}
+
+/// The result of performing a hit test on layers.
+class LayerHitTestResult<T> {
+  /// Creates an empty layer hit test result.
+  LayerHitTestResult()
+     : _path = <LayerHitTestEntry<T>>[];
+
+  /// An unmodifiable list of [LayerHitTestEntry] objects recorded during the
+  /// hit test, in the visual order from front to behind.
+  ///
+  /// The path starts with the most specific and visually front-most entry,
+  /// typically the one at the leaf of the layer tree and the one added the
+  /// latest among its siblings, then proceeds onto the visually behind and less
+  /// specific entries.
+  Iterable<LayerHitTestEntry<T>> get path => _path;
+  final List<LayerHitTestEntry<T>> _path;
+
+  /// Add a [LayerHitTestEntry] to the path.
+  ///
+  /// The new entry is added at the end of the path, which means entries should
+  /// be added in order from visually behind to front.
+  void add(LayerHitTestEntry<T> entry) {
+    _path.add(entry);
+  }
+
+  @override
+  String toString() => 'LayerHitTestResult(${_path.isEmpty ? "<empty path>" : _path.join(", ")})';
+}
+
```

Comment:
do we have to use new classes here, as opposed to reusing the equivalent classes from the gestures library?

Reply:
I'd love to reuse classes, but `handleEvent` doesn't make sense in layers https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/gestures/hit_test.dart#L42

Reply:
fix that one too!

Reply:
maybe we should adjust the gestures library objects so that they do make sense... not sure exactly how to do it

Reply:
We can also just use a `List<S>` here, since we only care about one list of values. Although it's not as future-proof.

Reply:
After discussion we discovered that, the mouse events also need `localPosition`, which require that extra classes are needed to be the return value. See https://github.com/flutter/flutter/issues/38488#issuecomment-527688254